File: 4_leetcode/C++/LC_1116_sem.cpp
#include <semaphore>
#include <iostream>
#include <thread>

class ZeroEvenOdd {
private:
    int n;
    int tick;
    std::binary_semaphore sem_zero;
    std::binary_semaphore sem_odd;
    std::binary_semaphore sem_even;

public:
    ZeroEvenOdd(int n) : 
        n(n), 
        tick(1), 
        sem_zero(1), 
        sem_odd(0), 
        sem_even(0) {}

    void zero(function<void(int)> printNumber) {
        while(true) {
            sem_zero.acquire();

            if(tick>n) {
                sem_odd.release();
                sem_even.release();
                break;
            }

            printNumber(0);

            if(tick%2!=0) sem_odd.release();
            else sem_even.release();
        }
    }

    void even(function<void(int)> printNumber) {
        while(true) {
            sem_even.acquire();

            if(tick>n) break;

            printNumber(tick++);

            sem_zero.release();
        }
    }

    void odd(function<void(int)> printNumber) {
        while(true) {
            sem_odd.acquire();

            if(tick>n) break;

            printNumber(tick++);

            sem_zero.release();
        }
    }
};

File: 4_leetcode/C#/LC_1116_sem.cs
using System;
using System.Threading;

public class ZeroEvenOdd {
    private int n;

    private int tick;
    private SemaphoreSlim semZero;
    private SemaphoreSlim semOdd;
    private SemaphoreSlim semEven;
    
    public ZeroEvenOdd(int n) {
        this.n=n;
        this.tick=1;

        semZero =new SemaphoreSlim(1, 1);
        semOdd  =new SemaphoreSlim(0, 1);
        semEven =new SemaphoreSlim(0, 1);
    }

    // printNumber(x) outputs "x", where x is an integer.
    public void Zero(Action<int> printNumber) {
        while (true) {
            semZero.Wait();

            if (tick>n) {
                semOdd.Release();
                semEven.Release();
                break;
            }

            printNumber(0);

            if (tick%2!=0) semOdd.Release();
            else semEven.Release();
        }
    }

    public void Even(Action<int> printNumber) {
        while (true) {
            semEven.Wait();

            if (tick>n) break;

            printNumber(tick++);

            semZero.Release();
        }
    }

    public void Odd(Action<int> printNumber) {
        while (true) {
            semOdd.Wait();

            if (tick>n) break;

            printNumber(tick++);

            semZero.Release();
        }
    }
}